drivers/hmis/app/models/hmis/form/definition_validator.rb

```diff
@@ -10,23 +10,47 @@ def self.perform(...)
   end
 
   # @param [Hash] document is a form definition document {'item' => [{...}] }
-  # @param [Hash] valid_pick_lists validate that pick list references are in this set
-  def perform(document, valid_pick_lists: [])
-    @issues = []
-    all_ids = check_ids(document, valid_pick_lists.to_set)
+  # @param [role] role of the form as string ('INTAKE', 'CLIENT', etc). If not provided, HUD rule validation will not occur.
+  def perform(document, _role = nil)
+    @issues = HmisErrors::Errors.new
+
+    # Validate JSON shape against JSON Schema
+    check_json_schema(document)
+    # Check Link IDs
+    all_ids = check_ids(document)
+    # Check references
     check_references(document, all_ids)
-    # we have cyclic dependencies and they seem to be supported by the front-end
-    # check_cycles(dependencies)
-    @issues
+
+    # TODO: Check HUD requirements (requires 'role')
+
+    @issues.errors
   end
 
   protected
 
   def add_issue(msg)
-    @issues.push(msg)
+    # TODO: resolve more details (e.g. link_id, section, actual severity level)
+    @issues.add(:definition, full_message: msg, severity: :error)
+  end
+
+  def check_json_schema(document)
+    schema_errors = Hmis::Form::Definition.validate_schema(document)
+    schema_errors.each do |e|
+      # Try to figure out which Link ID the error is on
+      item_path = /(\/item\/[0-9]){1,}/.match(e.to_s)&.try(:[], 0)&.split('/')&.
+        compact_blank&.
+        map { |s| Integer(s, exception: false) || s }
+      link_id = document.dig(*item_path, 'link_id') if item_path
+      if link_id
```

Comment:
copied and slightly amended from JsonForms. This shouldn't happen with the form builder once its stable, but until then its helpful to see the link id with the error